src/write.rs

```diff
@@ -548,9 +611,61 @@ where
                     }
                 }
             }
+            Operation::Retain(ref key, ref predicate) => {
+                if let Some(e) = inner.data.get_mut(key) {
+                    let mut del = 0;
+
+                    let len = e.len();
+
+                    for i in 0..len {
+                        if !predicate.eval(unsafe { e.get_unchecked(i) }) {
+                            del += 1;
+                        } else {
+                            e.swap(i - del, i);
+                        }
+                    }
+
+                    // manually truncate and forget values so they aren't dropped.
+                    while e.len() > (len - del) {
```

Comment:
I think we can just truncate with `e.set_len()` here?

Reply:
That is a great idea!